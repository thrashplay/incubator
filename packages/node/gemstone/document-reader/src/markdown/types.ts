import MarkdownIt from 'markdown-it'

import { ContentBlock, Section, Table } from '../types'

export type Token = ReturnType<ReturnType<typeof MarkdownIt>['parse']>[number]

export type TokenProcessingContextSharedFields = {
  complete: boolean
  remainingTokens: Token[]
}

export type TokenProcessingContext<TState extends any = any> =
  TokenProcessingContextSharedFields &
  {
    [k in keyof TState]: TState[k]
  }

export type TokenStreamHandler<TContext extends TokenProcessingContext = TokenProcessingContext> =
  (context: TContext) => TContext

/**
 * A logical block of Markdown tokens that make up a single section of meaningful content in a document.
 * A content block may combine multiple tokens that, on their own, aren't meaningful (such as open and close
 * tokens around a container of text) and combine them into a higher-order group.
 *
 * Example content blocks could be headings, paragraphs, data tables, etc.
 */
export interface MarkdownContentBlock extends ContentBlock {
  /** Markdown tokens comprising this block */
  tokens: Token[]
}

export interface MarkdownSection extends Section, MarkdownContentBlock {
  body: MarkdownContentBlock
}

export type MarkdownTable = Table & MarkdownContentBlock
