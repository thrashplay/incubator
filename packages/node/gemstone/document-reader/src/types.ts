/**
 * A logical block of Markdown tokens that make up a single section of meaningful content in a document.
 * A content block may combine multiple tokens that, on their own, aren't meaningful (such as open and close
 * tokens around a container of text) and combine them into a higher-order group.
 * 
 * Example content blocks could be headings, paragraphs, data tables, etc.
 */
export interface ContentBlock {
  /** gets this block's content as an HTML string */
  getHtml: () => string

  /** gets this block's content as a plain text string */
  getText: () => string
}

/**
 * Content block representing a section in a document.
 * A section consists of a heading and body content. Sections can be nested inside each other to form a
 * hierarchical document structure.
 */
export interface Section extends ContentBlock {
  /** 
   * Block representing all of the section's body content, including sub-sections. 
   * The section heading is not included.
   **/
  body: ContentBlock

  /** child sections of this section */
  children: Section[]

  /**
   * The depth of this section.
   * The document root is depth '0', and each level of nesting for sections increases this value by 1
   */
  depth: number

  /** the heading block for this section */
  heading?: Heading

  /** the title of this section, which is derived from it's heading */
  title?: string
}

/**
 * Content block representing a Heading in a document.
 */
export interface Heading extends ContentBlock {
  /** the numeric level of the heading (1-6) */
  level: number

  /** the text contents of this heading */
  text: string
}

export interface Document {
  /** array of sections, in the order they appeared in the source */
  sections: Section[]
}