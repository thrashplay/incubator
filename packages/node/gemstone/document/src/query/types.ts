import { Document, DocumentNode, SectionNode } from '../types'

/** root of all queries against a document: either the document itself, or a node inside it */
export type QueryRoot = Document | DocumentNode

export type SectionQueryRoot = Document | SectionNode
