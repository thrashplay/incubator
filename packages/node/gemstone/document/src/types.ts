import { ValuesType } from 'utility-types'

/**
 * A logical block content that make up a single portion of a document.
 * Example content blocks could be headings, paragraphs, data tables, etc.
 */
export interface Content {
  /** returns the content as plain text */
  getText: () => string
}

/** Node structure for documents */
export type DocumentNodeType = 
  'generic' | 
  'heading' | 
  'section' | 
  'root' |
  'table' |
  'table-body' |
  'table-footer' |
  'table-header'

export interface BaseDocumentNode {
  children: DocumentNode[]
  content: Content
}

export interface HeadingNode extends BaseDocumentNode {
  /** the numeric level of the heading (i.e. 1-6) */
  level: number

  type: 'heading'
}

export interface TableNode extends BaseDocumentNode { 
  /** the table's body, if it has one */
  body?: DocumentNode

  /** the names of the columns in the table */
  columnNames: string[]

  /** 
   * The table's data, represented as an array of Records 
   * Each record holds the data for a single table row. The record's keys are the column names, and the
   * values are the string contents of the corresponding cells.
   **/
  data: Record<string, string>[]

  /** the table's footer, if it has one */
  footer?: DocumentNode

  /** the table's body, if it has one */
  header?: DocumentNode

  /** number of rows in this table */
  rowCount: number

  type: 'table'
}
export type TableRowData = ValuesType<TableNode['data']>

export interface SectionNode extends BaseDocumentNode {
  /** the heading block for this section */
  heading?: HeadingNode

  /**
   * The depth of this section.
   * The document root is depth '0', and each level of nesting for sections increases this value by 1
   */
  level: number
  
  /** collection of table nodes contained in this section */
  tables: TableNode[]

  /** the title of this section, which is derived from it's heading */
  title?: string

  type: 'section'
}

type SpecialNodes = HeadingNode | SectionNode | TableNode

export type DocumentNode = SpecialNodes |
(
  BaseDocumentNode & 
  {
    type: Omit<DocumentNodeType, SpecialNodes['type']>
  }
)

export interface Document {
  /** the document's root node */
  root: DocumentNode
}