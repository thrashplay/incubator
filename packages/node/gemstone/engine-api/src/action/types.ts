import { Either } from 'monet'

import { Dictionary } from '@thrashplay/gemstone-model'

import { Entity } from '../entity'
import { LogEntry } from '../log'
import { TransformationDescriptor } from '../transformation'
import { WorldState } from '../world'

/**
 * An Action represents an attempt by one Entity to change the state of another. The type of action indicates
 * the nature of the change being attempted, and different types will have different details to describe how
 * the action is performed -- it's intensity, effectiveness, nature, etc.
 *
 * Every action also has a 'source', which is the entity initiating the action, and a 'target', which is the
 * entity being acted upon.
 */
export type Action<
  TAction extends string = string,
  TDetails extends Dictionary<string, unknown> = never
> = {
  source: Entity['id']
  target: Entity['id']
  type: TAction
} & ([TDetails] extends [never] ? unknown : { details: TDetails })

/** The widest possible type for an object we know is some kind of action. */
export type AnyAction = {
  details?: unknown
  source: Entity['id']
  target: Entity['id']
  type: Action['type']
}

/** Error type for failed Actions */
export type ActionError = LogEntry

/** Result data structure for action handlers. */
export type ActionResult = {
  /** Actions to take in response to the original action. */
  reactions?: Action | readonly Action[]

  /** Transformations to apply to the world state. */
  transformations?: TransformationDescriptor<string, any> | readonly TransformationDescriptor<string, any>[]
}

/**
 * Signature for ActionHandlerFunctions, which handle an action in the context of the a current world state.
 *
 * An ActionHandlerFunction is given the action to handle, as well as the world state. It is responsible
 * for generating an action result, which is a set of zero or more reactions (themselves Actions) to take
 * in response, and zero or more transformations. Transformations are ultimately responsible for applying
 * updates to the world state.
 */
export type ActionHandlerFunction<TAction extends AnyAction = AnyAction> = (
  action: TAction,
  world: WorldState
) => Either<ActionError, ActionResult>

/**
 * An action handler exposes two functions:
 *
 *  - 'supports', which is used to determine if the handler knows how to handle the specified action
 *  - 'handle', which is the ActionHandlerFunction that will handle supported actions
 *
 */
export type ActionHandler<TAction extends AnyAction = any> = {
  handle: (action: TAction, world: WorldState) => Either<LogEntry, ActionResult>
  supports: (action: AnyAction) => boolean
}
