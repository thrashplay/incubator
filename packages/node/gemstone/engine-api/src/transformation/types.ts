import { Either } from 'monet'

import { LogEntry } from '../log'
import { NthArgOrNever, OptionalRestParameter } from '../types'
import { WorldState } from '../world'

/**
 * A TransformationDescriptor encodes a parameterized change to the simulation's World state. It has a string
 * representing the type of transformation, and an optional type-specific parameter value.
 */
export type TransformationDescriptor<
  TType extends string = string,
  TParameter extends any = never
> = {
  type: TType
} & ([TParameter] extends [never] ? unknown : { parameter: TParameter })

/** Error type for failed Transformations */
export type TransformationError = LogEntry

/** Signature for transformation functions that return Either an error, or updated world state. */
export type TransformationFunction<TParameter extends any = never> = (
  world: WorldState,
  ...parameters: OptionalRestParameter<TParameter>
) => Either<TransformationError, WorldState>

/** Signature for transformation functions that will never fail, and always return an updated world state. */
export type UniversalTransformationFunction <TParameter extends any = never> = (
  world: WorldState,
  ...parameters: OptionalRestParameter<TParameter>
) => WorldState

export type TransformationSet = Readonly<{
  [k in string]: TransformationFunction<any> | UniversalTransformationFunction<any>
}>

/**
 * The union of all possible Transformation shapes handled by a collection of TransformationFunctions.
 * Each type in this union, which is a type of 'Transformation' has the following characteristics
 *
 *   - its "TType" is a key in the set of transformation functions
 *   - its "TParameter" is the type of parameter taken by the corresponding transformation function, or
 *     'never' if there isn't one
 *
 **/
export type TransformationTypes<
  TTransformations extends TransformationSet = any
> = ({
  [K in keyof TTransformations]: TTransformations[K] extends (
    world: WorldState, ...rest: any[]
  ) => ReturnType<TransformationFunction> | ReturnType<UniversalTransformationFunction>
    ? TransformationDescriptor<K & string, NthArgOrNever<TTransformations[K], 1>>
    : TransformationDescriptor<K & string, unknown>
})[keyof TTransformations]
