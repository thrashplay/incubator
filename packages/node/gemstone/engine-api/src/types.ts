// TODO: is T | never actually valid, compared to optional/undefined
export type Dictionary<K extends string | number | symbol = string, T = unknown> = { [k in K]: T | never }

export type OptionalRestParameter<TType> = [TType] extends [never] ? [] : [TType]

export type AnyFunction = (...args: any[]) => any
