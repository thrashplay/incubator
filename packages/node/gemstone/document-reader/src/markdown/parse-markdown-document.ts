import MarkdownIt from 'markdown-it'
import { mapAt } from '@thrashplay/fp'
import { findLastIndex, flow } from 'lodash/fp'

import { Document } from '../types'

import { MarkdownSection, TokenProcessingContext } from './types'
import { parseSection } from './parse-section'
import { createTokenProcessor } from './create-token-processor'


type DocumentProcessingContext = TokenProcessingContext<{
  /** sections that have already been successfully parsed */
  sections: MarkdownSection[]
}>

/**
 * Adds a section as a child to the most recently parsed section with a depth that is lower (that is, a higher 
 * level section in the document hierarchy). Does not modify any sections if there is no higher-level section.
 */
const addSectionToParent = (section: MarkdownSection) => (context: DocumentProcessingContext) => {
  const depthLessThan = (section: MarkdownSection) => (item: MarkdownSection) => item.depth < section.depth
  const addChild = (child: MarkdownSection) => (parent: MarkdownSection) => ({
    ...parent,
    children: [...parent.children, child],
  })

  const { sections } = context
  const parentIndex = findLastIndex(depthLessThan(section))(sections)
  return parentIndex === -1 ? context : {
    ...context,
    sections: mapAt(parentIndex, addChild(section))(sections),
  }
}

/**
 * Adds a section to the 'sections' array in the document.
 */
const addSectionToDocument = (section: MarkdownSection) => (context: DocumentProcessingContext) => {
  return {
    ...context,
    sections: [...(context.sections ?? []), section],
  }
}

const parseNextSection = (context: DocumentProcessingContext) => {
  const { currentSection, remainingTokens } = parseSection(context.remainingTokens)
  return flow(
    addSectionToParent(currentSection),
    addSectionToDocument(currentSection),
    (context: DocumentProcessingContext) => ({
      ...context,
      remainingTokens,
    }),
  )(context)
}

const parseDocument = createTokenProcessor(
  [
    parseNextSection,
  ],
  () => ({ sections: [] }),
)

export const parseMarkdownDocument = (content: string): Promise<Document> => {
  return new Promise<Document>((resolve: (document: Document) => void) => {
    const markdown = new MarkdownIt()
    const tokens = markdown.parse(content, {})

    resolve(parseDocument(tokens))
  })
}
