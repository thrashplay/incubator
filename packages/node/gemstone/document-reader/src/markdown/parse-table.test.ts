import fs from 'fs'
import path from 'path'

import { mapValues } from 'lodash/fp'

import { ContentBlock, Document, Section } from '../types'

import { parseMarkdownDocument } from './parse-markdown-document'

const readFixture = (fixtureName: string) => {
  return fs.readFileSync(path.resolve(__dirname, '__fixtures__', 'tables', fixtureName), 'utf8')
}

const doParse = (fixtureName: string) => {
  return parseMarkdownDocument(readFixture(fixtureName))
}

describe('markdown table parsing', () => {
  describe('section content', () => {
    let document: Document
    let section: Section
    
    beforeEach(async () => {
      document = await doParse('with-other-content.md')
      expect(document.sections.length).toBe(1)
      section = document.sections[0]
    })

    it('has correct HTML representation', () => {
      expect(section.getHtml()).toMatchSnapshot()
    })

    it('has correct plain text representation', () => {
      expect(section.getText()).toMatchSnapshot()
    })
  })

  describe('with no other content', () => {
    let document: Document
    let section: Section
    
    beforeEach(async () => {
      document = await doParse('with-no-other-content.md')
      expect(document.sections.length).toBe(1)
      section = document.sections[0]
    })

    it('parses table', () => {
      expect(section.tables.length).toBe(1)
    })

    it('has correct number of rows', () => {
      const table = section.tables[0]
      expect(table.rows).toBe(2)
    })

    it('has correct column names', () => {
      const table = section.tables[0]
      expect(table.columnNames).toStrictEqual([
        'Column A',
        'B',
        'Last',
      ])
    })

    it('has correct HTML representation', () => {
      const table = section.tables[0]
      expect(table.getHtml()).toMatchSnapshot()
    })

    it('has correct plain text representation', () => {
      const table = section.tables[0]
      expect(table.getText()).toMatchSnapshot()
    })

    it.each<[number, Record<string, string>]>([
      [0, { 'Column A': 'Two bold words', 'B': '2', 'Last': 'Lorem' }],
      [1, { 'Column A': 'Row 2', 'B': '4', 'Last': 'Ipsum' }],
    ])('has correct data in row: %p', (row, expectedData) => {
      const table = section.tables[0]
      expect(table.data[row]).toStrictEqual(expectedData)
    })

    it.each<[number]>([
      [0],
      [1],
    ])('has correct cell contents in row: %p', (row) => {
      const table = section.tables[0]
      const dataAsHtml = mapValues((block: ContentBlock) => block.getHtml())(table.rowContent[row])
      expect(dataAsHtml).toMatchSnapshot()
    })
  })
})