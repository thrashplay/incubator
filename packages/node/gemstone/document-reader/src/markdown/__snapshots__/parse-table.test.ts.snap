// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown table parsing section content has correct HTML representation 1`] = `
"<h1>Section with Table</h1>
<p>Lorem ipsum dolor sit amet.</p>
<table>
<thead>
<tr>
<th>Column A</th>
<th>B</th>
<th>Last</th>
</tr>
</thead>
<tbody>
<tr>
<td>Two word</td>
<td>2</td>
<td>Lorem</td>
</tr>
<tr>
<td>Row 2</td>
<td>4</td>
<td>Ipsum</td>
</tr>
</tbody>
</table>
<p>Consectitur, adspicing elit.</p>"
`;

exports[`markdown table parsing section content has correct plain text representation 1`] = `
"Section with Table
Lorem ipsum dolor sit amet.

Column A   B   Last    
Two word   2   Lorem   
Row 2      4   Ipsum   

Consectitur, adspicing elit."
`;

exports[`markdown table parsing with no other content has correct HTML representation 1`] = `
"<table>
<thead>
<tr>
<th>Column A</th>
<th>B</th>
<th>Last</th>
</tr>
</thead>
<tbody>
<tr>
<td>Two <strong>bold</strong> words</td>
<td>2</td>
<td>Lorem</td>
</tr>
<tr>
<td>Row 2</td>
<td>4</td>
<td>Ipsum</td>
</tr>
</tbody>
</table>"
`;

exports[`markdown table parsing with no other content has correct cell contents in row: 0 1`] = `
Object {
  "B": "<td>2</td>",
  "Column A": "<td>Two <strong>bold</strong> words</td>",
  "Last": "<td>Lorem</td>",
}
`;

exports[`markdown table parsing with no other content has correct cell contents in row: 1 1`] = `
Object {
  "B": "<td>4</td>",
  "Column A": "<td>Row 2</td>",
  "Last": "<td>Ipsum</td>",
}
`;

exports[`markdown table parsing with no other content has correct plain text representation 1`] = `
"Column A         B   Last    
Two bold words   2   Lorem   
Row 2            4   Ipsum"
`;
